src/main/scala/com/gu/invoicing/refund/Impl.scala

```diff
@@ -0,0 +1,246 @@
+   * by how much by taking into account any previous adjustments already made to corresponding items.
+   */
+  def spreadRefundAcrossItems(
+    invoiceItems: List[InvoiceItem],
+    adjustments: List[InvoiceItemAdjustment],
+    totalRefundAmount: Double,
+    refundGuid: String,
+  ): List[InvoiceItemAdjustmentWrite] = {
+
+    /* Collect all item adjustments of a particular invoice item and return remaining amount that can be adjusted/refunded */
+    def availableAmount(invoiceItem: InvoiceItem): Option[Double] = {
+      netAdjustmentsByInvoiceItemId(adjustments).get(invoiceItem.Id) match {
+        case Some(netAdjustment) =>
+          val availableRefundableAmount = invoiceItem.ChargeAmount - netAdjustment
+          if (availableRefundableAmount <= 0.0) None else Some(availableRefundableAmount)
+
+        case None => // this items has not been adjusted therefore the original full item amount is available
+          Some(invoiceItem.ChargeAmount)
+      }
+    }
+
+    @tailrec def loop(remainingAmounToRefund: Double, remainingItems: List[InvoiceItem], accumulatedAdjustments: List[InvoiceItemAdjustmentWrite]): List[InvoiceItemAdjustmentWrite] = {
+      remainingItems match {
+        case Nil =>
+          accumulatedAdjustments
+
+        case nextItem :: tail =>
+          val adjustItemBy: Double => InvoiceItemAdjustmentWrite =
+            InvoiceItemAdjustmentWrite(LocalDate.now(), _, refundGuid, nextItem.InvoiceId, "Credit", "InvoiceDetail", nextItem.Id)
+
+          availableAmount(nextItem) match {
+            case Some(availableRefundableAmount) =>
+              if ((remainingAmounToRefund - availableRefundableAmount) <= 0.0)
+                adjustItemBy(remainingAmounToRefund) :: accumulatedAdjustments
+              else
+                loop(remainingAmounToRefund - availableRefundableAmount, tail, adjustItemBy(availableRefundableAmount) :: accumulatedAdjustments)
+
+            case None =>
+              loop(remainingAmounToRefund, tail, accumulatedAdjustments)
+          }
+
+      }
+    }
+
+    loop(totalRefundAmount, invoiceItems, List.empty[InvoiceItemAdjustmentWrite])
+  }
+
+  def applyRefundOverItemAdjustments(invoiceItems: List[InvoiceItemAdjustmentWrite]): List[AdjustmentResult] = {
+    HttpWithLongTimeout(s"$zuoraApiHost/v1/action/create")
+      .header("Authorization", s"Bearer ${accessToken()}")
+      .header("Content-Type", "application/json")
+      .postData(write(InvoiceItemAdjustmentsWriteRequest(objects = invoiceItems, `type` = "InvoiceItemAdjustment")))
+      .method("POST")
+      .asString
+      .body
+      .pipe(read[List[AdjustmentResult]](_))
+  }
+
+  def joinInvoiceWithInvoiceItemsOnInvoiceIdKey(invoices: List[Invoice], invoiceItemsBySubscription: Map[String, List[InvoiceItem]]): List[(String, Invoice, List[InvoiceItem])] = {
+    for {
+      (invoiceId, invoice) <- invoices.map(invoice => invoice.Id -> invoice)
+      (`invoiceId`, invoiceItems) <- invoiceItemsBySubscription
```

Comment:
Confusing

Reply:
Indeed, thanks! These are my favourite review comments as they force me to find a simpler solution.